File: src/com/personal/util/TreeNode.java
package com.personal.util;

import java.io.Serializable;

/**
 * Created by prajeev on 19/3/17.
 */
public class TreeNode implements Serializable{
    public ComparableNumber number;
    public Entry<Integer, Integer> entry;
    public String data;
    public TreeNode left;
    public TreeNode right;


    public TreeNode(int number){
        ComparableNumber comparableNumber = new ComparableNumber(number);
        this.number = comparableNumber;
    }

    public TreeNode(String data) {
        this.data = data;
    }

    @Override
    public boolean equals(Object obj) {

        if(obj instanceof TreeNode){
            TreeNode that = (TreeNode) obj;
            if(this.number.equals(that.number)){
                return true;
            }
        }
        return false;
    }

    @Override
    public String toString() {
        if(number!=null)
            return number.toString();
        else
            return data;
    }

    public static TreeNode merge(TreeNode thiz, TreeNode that) {
        if(thiz == null) return that;
        if(that == null) return thiz;
        TreeNode node = new TreeNode(thiz.number.number + that.number.number);
        node.left = merge(thiz.left , that.left);
        node.right = merge(thiz.right , that.right);
        return node;
    }
}

File: src/com/functional/util/TreeNode.scala
package com.functional.util

/**
  * Created by prajeeva on 9/7/17.
  */
case class TreeNode(var data:Int , var left:TreeNode,var right:TreeNode) {
  override def canEqual(that: Any): Boolean = {
    if(that.isInstanceOf[TreeNode]) return true
    false
  }

  override def equals(o: scala.Any): Boolean = {
    if(this.canEqual(o)) {
      var that = o.asInstanceOf[TreeNode]
      if(this.data != that.data) {
        return false
      }
      if(this.left!=null && !this.left.equals(that.left)) {
        return false
      }
      if(this.right!=null && !this.right.equals(that.right)) {
        return false
      }
    }
    true
  }
}

object TreeNode {
  def apply: TreeNode = new TreeNode(0,null,null)
  def apply(data:Int):TreeNode = new TreeNode(data,null,null)

  def +(x:TreeNode , y:TreeNode):TreeNode = {
    if(x == null) return y
    if(y == null) return x
    this.apply
  }
}
